### PR TITLE
Features/provide field extensions

### PR DESCRIPTION
**Pull request by rkorzeniec**
During an upgrade of the [graphql](https://github.com/rmosolgo/graphql-ruby) gem from `1.9.x` to `1.10.x` it became apparent that `Field Instrumentation` is no longer supported (extra info [here](https://github.com/rmosolgo/graphql-ruby/blob/master/guides/queries/interpreter.md#compatibility)) and that the `Field Extension` is the [supported replacement](https://github.com/rmosolgo/graphql-ruby/issues/2903#issuecomment-619391149) for that.

This PR introduces `Field Extension` API, which is based on the implementation of the `Field Instrument` interface. 

Further Field Extension documentation: https://graphql-ruby.org/type_definitions/field_extensions.html

### CODE REVIEW

**Comment by rkorzeniec** [comment]
@severin I am opening the PR also on our fork, to get the code better and not be blocked by the maintainer of the gem. So that on our end we can proceed with the GQL upgrade. 

Once this is merged I can switch it on our main app, then we can wait until the maintainer of the gem responds, then we can make the switch back.

**Comment by severin** [comment]
> @severin I am opening the PR also on our fork, to get the code better and not be blocked by the maintainer of the gem. So that on our end we can proceed with the GQL upgrade.
> 
> Once this is merged I can switch it on our main app, then we can wait until the maintainer of the gem responds, then we can make the switch back.

👍

File: spec/graphql/result_cache/key_spec.rb
Comment (severin) [inline spec/graphql/result_cache/key_spec.rb:27]: Wouldn't `it "should not include path"` be more clear?

**Review by severin — COMMENTED** [review]
@rkorzeniec the code changes look good 👍 

Did you/could you push a branch of https://github.com/onrunning/on that uses this version of the result cache gem so that I could test it in practice?

**Comment by rkorzeniec** [comment]
@severin https://github.com/onrunning/on/tree/updates/graphql-field-extensions

File: spec/graphql/result_cache/key_spec.rb
Comment (rkorzeniec) [reply spec/graphql/result_cache/key_spec.rb:27]: 💯

File: lib/graphql/result_cache/key.rb
Comment (severin) [inline lib/graphql/result_cache/key.rb]: This could lead to cache key collisions in the (very contrived) example of adding result caching to nested fields with the same name on different types. It's a non-issue for us at the moment since we only use result caching on top-level fields, but it's a little timebomb waiting for that contrived example to actually happen 😉 

Example: the Filter Page type has a field `topBanner`. If you were to apply result caching to only this field (and remove the enclosing result caching on the whole `filterPage` field on the query type) then `field_clause == "topBanner"`. This could lead to a collision if for example the Shoe Finder type also has a field called `topBanner` that is also result cached.

Luckily, `@field.path` also exists and returns the full path (in this case `"filterPage.topBanner"`). So I would use `path` instead of `name` in the `field_clause`.

Here is the code to reproduce my example from above:

```ruby
# in Types::FilterPage::PageType:

  top_banner_result_cache_config = On::Graphql::ResultCache::Config.call(:top_banner)
  field :top_banner, Types::FilterPage::TopBannerType, null: true do
    extension GraphQL::ResultCache::FieldExtension, top_banner_result_cache_config
  end

  def top_banner(*args)
    object.top_banner
  end

# in Types::QueryType -> remove the result cache extension for the `filter_page` field
```

Then test against http://localhost:3001/collection/outdoor

File: lib/graphql/result_cache/callback.rb
Comment (severin) [inline lib/graphql/result_cache/callback.rb]: Same would apply here: `@field.path` should be used over `@field.name`. Though it does not really matter since it only affects logging...

**Review by severin — CHANGES_REQUESTED** [review]
@rkorzeniec works great when using the main app pointed to this branch 👍 Great job on finding this solution and on the clean implementation 👏 

I found a tiny edge case issue that I think should be addressed, see comment(s). What do you think?

File: lib/graphql/result_cache/key.rb
Comment (rkorzeniec) [reply lib/graphql/result_cache/key.rb]: Yes! I agree, thank you for catching that case 👍

File: lib/graphql/result_cache/callback.rb
Comment (rkorzeniec) [reply lib/graphql/result_cache/callback.rb]: 💯

File: lib/graphql/result_cache/key.rb
Comment (rkorzeniec) [reply lib/graphql/result_cache/key.rb]: After recent push this works.
miss:
```ruby
[28, 37] in /Users/robertkorzeniec/.rbenv/versions/2.5.8/lib/ruby/gems/2.5.0/bundler/gems/graphql-result_cache-c2ae50c5f78a/lib/graphql/result_cache/key.rb
   28:         @ctx.path.join('.') unless @ctx.nil?
   29:       end
   30:
   31:       def field_clause
   32:         byebug
=> 33:         @field.path unless @field.nil?
   34:       end
   35:
   36:       def args_clause
   37:         @args.to_h.to_a.flatten
(byebug) @field.path
"FilterPage.topBanner"
(byebug) c
GraphQL result cache key miss: GraphQL:Result::FilterPage.topBanner:en/ch/ba5b0caee7b5bc7198c8dd0108a5a52b39c68069ace2aaa36f430b79b65f0cc6
```
hit:
```ruby
[28, 37] in /Users/robertkorzeniec/.rbenv/versions/2.5.8/lib/ruby/gems/2.5.0/bundler/gems/graphql-result_cache-c2ae50c5f78a/lib/graphql/result_cache/key.rb
   28:         @ctx.path.join('.') unless @ctx.nil?
   29:       end
   30:
   31:       def field_clause
   32:         byebug
=> 33:         @field.path unless @field.nil?
   34:       end
   35:
   36:       def args_clause
   37:         @args.to_h.to_a.flatten
(byebug) @field.path
"FilterPage.topBanner"
(byebug) c
GraphQL result cache key hit: GraphQL:Result::FilterPage.topBanner:en/ch/ba5b0caee7b5bc7198c8dd0108a5a52b39c68069ace2aaa36f430b79b65f0cc6
```

**Review by severin — APPROVED** [review]
👌

**Pull request merged**
